Fix potential to miss file download sequences

benletchford opened the pull request:
Hi @butlerx (and @kb1ooo)

I've changed the way the file buffering works and it now builds the file buffer character by character not caring how many data frames the data is split up upon.

This fixes issue #214. 
* This now also allows you to receive multiple files in the one data frame (which the current method doesn't support). 
* This method also correctly returns data to the left and right side of a data frame if it's not part of a file or file sequences.

Lastly, I've now made the download feature opt-in by passing `--enablefiledownload` to the node server which passes it to the client webpage (this also opens up the possibility to pass other config to the client, eg styling from the server-side [loosely referenced by #216]).

I'm open to the idea of splitting these two things into different PRs but I thought they go very nicely together.

Let me know your thoughts on this :)

kb1ooo commented:
Looks good upon a cursory review.  I think the only issue, which is probably a non issue, is if lets say 3 of the FILE_BEGIN characters are intercepted (and they are not meant to delineate a file), they won't get written to the terminal until a 4th character is intercepted.  Which may not happen or not happen for some time.

benletchford commented:
> Looks good upon a cursory review. I think the only issue, which is probably a non issue, is if lets say 3 of the FILE_BEGIN characters are intercepted (and they are not meant to delineate a file), they won't get written to the terminal until a 4th character is intercepted. Which may not happen or not happen for some time.

You're spot on here. However, I think this may be unavoidable. We can't send it to the terminal until we know for sure that it *isn't* a `FILE_BEGIN` sequence.

I also saw your comment in regard to the other issues you found. I found the same problems while thinking through this which ultimately led to the decision to buffer character by character (simplifies things a lot).

> 1. Data in the buffer, but outside the begin/end delimiters, is lost (i.e. not written to the terminal).
> 2. A FILE_BEGIN sequence that occurs after a FILE_BEGIN but before a FILE_END is not handled properly. 
> 3. Or similarly a FILE_END that occurs before a FILE_BEGIN has occurred is not ignored.
There may be multiple segments that occur within a single buffer and that is not handled.

All these cases should are now handled correctly now that I'm buffering character by character.

Once again, thanks for your insights.

kb1ooo commented:
> You're spot on here. However, I think this may be unavoidable. We can't send it to the terminal until we know for sure that it _isn't_ a `FILE_BEGIN` sequence.
> 

I think I agree.  I wasn't sure what is the lessor of the 2 evils, writing the extra chars to the terminal or not writing.  Attached is a parser I wrote (abstracted from wetty) that takes the former approach.  It just prepends the current buffer with delim.length - 1 characters from the previous buffer, recurses on looking for beg/end, and then fixes the file buffer if extra characters had been written to it previously.  So file buffers always come out clean but terminal write may not.   I'm not a javascript guy, but it was a decent exercise given that I've started to have to help my high school daughter with javascript :)

```javascript
//var FILE_BEGIN = '\u001b[5i';
//var FILE_END = '\u001b[4i';
var FILE_BEGIN = '[[[';
var FILE_END = ']]]';
var delimSize = FILE_BEGIN.length;
var dataWrite = "";
var dataFile = "";

function onComplete() {
    console.log("FILE: " + dataFile);
    dataFile = "";
}

function parseEnd(data, aug) {
    dataAug = aug.concat(data);
    indexOfFileEnd = dataAug.indexOf(FILE_END);

    if(indexOfFileEnd != -1) {
        if(indexOfFileEnd > aug.length) // add characters less the augmented piece
            dataFile = dataFile.concat(dataAug.substring(aug.length, indexOfFileEnd));
        else if(dataFile.length > 0)    // remove marker characters that were written already
            dataFile = dataFile.substring(0, dataFile.length - aug.length + indexOfFileEnd);

        // have a full file
        onComplete();
        
        // recurse if there's any data left in the buffer
        if(indexOfFileEnd + FILE_END.length < dataAug.length) 
            parseBeg(dataAug.substring(indexOfFileEnd + FILE_END.length, dataAug.length),"");
    }
    else {
        dataFile = dataFile.concat(data);
    }
}

function parseBeg(data, aug) {

    dataAug = aug.concat(data);
    indexOfFileBegin = dataAug.indexOf(FILE_BEGIN);

    if(dataFile.length == 0 && indexOfFileBegin != -1) {
        if(indexOfFileBegin > aug.length) // write any data up to the begin marker
            dataWrite = dataWrite.concat(dataAug.substring(aug.length, indexOfFileBegin));
        if(indexOfFileBegin + FILE_BEGIN.length < dataAug.length) // look for end marker
            parseEnd(dataAug.substring(indexOfFileBegin + FILE_BEGIN.length, dataAug.length), "");
    }
    else if(dataFile.length > 0) { // look for end marker
        parseEnd(data, aug);
    }
    else {
        dataWrite = dataWrite.concat(data);
    }
}

function returnTail(data) {
    return data.substring(data.length - delimSize + 1, data.length);
}

//
// string is "[[[he[[[llo]]]the]]]re[[[dude]]]monkey"
//
// Correct response should be:
//
// FILE: he[[[llo
// FILE: dude
// WRITE: the]]]remonkey
//
// Note that when the buffer is split, a partial begin seg is written and you get the following instead
//
// FILE: he[[[llo
// FILE: dude
// WRITE: [[the]]]remonkey
//

var one_buf = FILE_BEGIN + "he" + FILE_BEGIN + "llo" + FILE_END + "the" + FILE_END + "re" + FILE_BEGIN + "dude" + FILE_END + "monkey";

console.log("Full buffer is " + one_buf);

console.log("Parsing as one buffer...");
console.log();

parseBeg(one_buf,"");
console.log("WRITE: " + dataWrite);
dataWrite = "";

console.log();
console.log("Parsing with splits across three buffers...");
console.log();
var first_buf = "[["
var second_buf = "[" + "he" + FILE_BEGIN + "llo" + FILE_END + "the" + FILE_END + "re" + FILE_BEGIN + "dude" + "]]"
var third_buf = "]" + "monkey";
parseBeg(first_buf, "");
parseBeg(second_buf, returnTail(first_buf));
parseBeg(third_buf, returnTail(second_buf));

console.log("WRITE: " + dataWrite);
```

benletchford commented:
@butlerx, had an opportunity to look at this? I wrote a bunch of unit tests for this as well but deliberately didn't include them as I didn't want to introduce a testing framework.

butlerx commented on src/client/index.ts:
```diff
@@ -14,6 +16,7 @@ import loadOptions from './options';
 import { copySelected, copyShortcut } from './copyToClipboard';
 import './wetty.scss';
 import './favicon.ico';
+import startWeTTy from '../server/wetty';
```
I'd like to avoid having the client import from the server

benletchford replied:
That's a mistake, thanks.

butlerx commented on src/server/cli/index.ts:
```diff
@@ -6,8 +6,8 @@ import { unWrapArgs } from './parseArgs';
-    const { ssh, server, command, ssl } = unWrapArgs(opts);
-    WeTTy(ssh, server, command, ssl).catch(err => {
+    const { ssh, server, client, command, ssl } = unWrapArgs(opts);
+    WeTTy(ssh, server, client, command, ssl).catch(err => {
```
I believe this would be a breaking change to the api.
Either update the package.json to reflect this or make this arg optional. 

Please also update the docs aswell

butlerx commented on src/server/socketServer/html.ts:
```diff
@@ -13,6 +14,9 @@ export default (base: string, title: string) => (
     <title>${title}</title>
     <link rel="stylesheet" href="${resourcePath}public/index.css" />
+    <script>
+      window.wettyConfig = ${JSON.stringify(clientConf)};
```
Configuration of the client should be done using localstorage https://github.com/butlerx/wetty/blob/master/src/client/options.ts

butlerx commented:
Please feel free to add test. If possible could you add a github action to run those tests on every pr

Commit:
buffer files on a character basis

Commit:
Add tests and initial github workflow

Commit:
Bump version

benletchford commented:
Hi @butlerx, I've tidied this up and removed my `--enablefiledownload` feature, I may add that as a new PR once this is through.

I've also added tests and setup a github action to run on every push / PR.

benletchford commented:
Hi @butlerx, any chance to review again?

butlerx commented on src/client/download.ts:
```diff
@@ -1,55 +1,100 @@
+    ) {
+      return data;
+    }
+    let newData = '';
```
can we replace this with
 ``` js
return data.map(this.bufferCharacter).join('')
```
i think it looks cleaner

benletchford replied:
I think this would actually have to be `return data.split('').map(this.bufferCharacter.bind(this)).join('')`?

butlerx replied:
yes youre right

butlerx reviewed (changes requested):
looks good just a stylistic change

butlerx commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,20 @@
+          node-version: 12
+      - run: yarn
+        name: Install dependencies
+      - run: yarn build
```
lets run a yarn lint on this as the code will need to pass lint to be published to npm.

benletchford replied:
Done.

butlerx commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,20 @@
+---
+on: [push, pull_request]
```
let's make this only on pull_request, there's a workflow for publishing to npm on push lest add the build step to that instead

benletchford replied:
Done.

Commit:
Pull request feedback from @butlerx

Commit:
ESlint ignore download.spec.ts file
